File: Kulenechki/Game/bot.js

const TelegramBot = require('node-telegram-bot-api');

// Замените 'YOUR_TELEGRAM_BOT_TOKEN' на токен вашего бота
const token = 'YOUR_TELEGRAM_BOT_TOKEN';

// Создаем бота
const bot = new TelegramBot(token, { polling: true });

// Константы для игры
const REQUIRED_MARKERS = 7;
const MAX_TENEHOD_POSITION = 5;

// Хранение данных пользователей в памяти
const users = {};

// Функция для отправки главного меню
function showMainMenu(chatId) {
  const options = {
    reply_markup: {
      inline_keyboard: [
        [{ text: '🚶‍♂️ Переместиться', callback_data: 'move' }],
        [{ text: '🔍 Искать маркеры', callback_data: 'search' }],
        [{ text: '🛠 Способность', callback_data: 'ability' }],
        [{ text: '📊 Статус', callback_data: 'status' }]
      ]
    }
  };
  bot.sendMessage(chatId, 'Выберите действие:', options);
}

// Обработчик команды /start
bot.onText(/\/start/, (msg) => {
  const chatId = msg.chat.id;
  const userId = msg.from.id;

  // Инициализируем данные пользователя
  users[userId] = {
    character: null,
    location: '🏡 Пушнор',
    markers: 0,
    tenehodPosition: 0,
    energy: 3,
    abilityUsed: false,
    specialItems: []
  };

  // Предлагаем выбрать персонажа
  const options = {
    reply_markup: {
      inline_keyboard: [
        [{ text: '🌟 Лумик', callback_data: 'character_Лумик' }],
        [{ text: '🎉 Флаффи', callback_data: 'character_Флаффи' }],
        [{ text: '🧠 Тинори', callback_data: 'character_Тинори' }],
        [{ text: '🌈 Ириска', callback_data: 'character_Ириска' }]
      ]
    }
  };

  bot.sendMessage(chatId, `Привет, ${msg.from.first_name}! Выберите своего Куленёчка:`, options);
});

// Обработчик нажатий на кнопки
bot.on('callback_query', (callbackQuery) => {
  const msg = callbackQuery.message;
  const data = callbackQuery.data;
  const userId = callbackQuery.from.id;
  const user = users[userId];

  if (!user) {
    bot.sendMessage(msg.chat.id, 'Пожалуйста, начните игру с команды /start.');
    return;
  }

  // Обработка выбора персонажа
  if (data.startsWith('character_')) {
    const character = data.replace('character_', '');
    user.character = character;
    bot.sendMessage(msg.chat.id, `Вы выбрали ${character}! Приключение начинается!`);
    showMainMenu(msg.chat.id);
  }

  // Обработка главного меню
  else if (data === 'move') {
    chooseLocation(msg.chat.id, user);
  } else if (data === 'search') {
    search(msg.chat.id, user);
  } else if (data === 'status') {
    showStatus(msg.chat.id, user);
  } else if (data === 'ability') {
    useAbility(msg.chat.id, user);
  }

  // Обработка выбора локации
  else if (data.startsWith('location_')) {
    const newLocation = data.replace('location_', '');
    user.location = newLocation;
    bot.sendMessage(msg.chat.id, `Вы переместились в ${newLocation}.`);
    showMainMenu(msg.chat.id);
  }
});

// Функция выбора локации
function chooseLocation(chatId, user) {
  const currentLocation = user.location;
  const locations = {
    '🏡 Пушнор': ['🌲 Светящаяся поляна', '🌫 Туманная роща'],
    '🌲 Светящаяся поляна': ['🏡 Пушнор', '🌊 Ручей Забытых Снов', '💧 Магический водопад'],
    '🌫 Туманная роща': ['🏡 Пушнор', '🕯 Таинственная пещера'],
    '💧 Магический водопад': ['🌲 Светящаяся поляна'],
    '🕯 Таинственная пещера': ['🌫 Туманная роща'],
    '🌊 Ручей Забытых Снов': ['🌲 Светящаяся поляна', '🦋 Поле светлячков'],
    '🦋 Поле светлячков': ['🌊 Ручей Забытых Снов']
  };
  const availableLocations = locations[currentLocation] || [];

  if (availableLocations.length > 0) {
    const keyboard = availableLocations.map(loc => [{ text: loc, callback_data: `location_${loc}` }]);
    const options = {
      reply_markup: {
        inline_keyboard: keyboard
      }
    };
    bot.sendMessage(chatId, 'Куда вы хотите переместиться?', options);
  } else {
    bot.sendMessage(chatId, 'Нет доступных локаций для перемещения.');
    showMainMenu(chatId);
  }
}

// Функция поиска маркеров
function search(chatId, user) {
  const events = ['positive', 'negative', 'special'];
  const event = events[Math.floor(Math.random() * events.length)];

  if (event === 'positive') {
    user.markers += 1;
    user.energy = Math.min(user.energy + 1, 3); // Восстанавливаем энергию
    bot.sendMessage(chatId, "🎉 Вы нашли маркер 'Света Дружбы' и восстановили 1 энергию!");
    checkVictoryCondition(chatId, user);
  } else if (event === 'negative') {
    advanceTenehod(chatId, user);
  } else {
    specialEvent(chatId, user);
  }
}

// Функция специальных событий
function specialEvent(chatId, user) {
  const specialEvents = ['hedgehog', 'trap', 'artifact'];
  const event = specialEvents[Math.floor(Math.random() * specialEvents.length)];

  if (event === 'hedgehog') {
    user.energy = Math.min(user.energy + 1, 3);
    bot.sendMessage(chatId, '🦔 Вы встретили Ежион! Он восстановил вашу энергию на 1 единицу.');
  } else if (event === 'trap') {
    user.tenehodPosition += 1;
    bot.sendMessage(chatId, '🕸 Вы попали в теневую ловушку! Тенеход продвинулся вперёд.');
  } else if (event === 'artifact') {
    user.specialItems.push('✨ Волшебный артефакт');
    bot.sendMessage(chatId, '✨ Вы нашли Волшебный артефакт! Теперь вы собираете на 1 маркер больше.');
  }
  showMainMenu(chatId);
}

// Функция использования способности
function useAbility(chatId, user) {
  if (user.energy <= 0) {
    bot.sendMessage(chatId, '⚠️ У вас недостаточно энергии для использования способности.');
    showMainMenu(chatId);
    return;
  }

  if (user.abilityUsed && user.character === '🌈 Ириска') {
    bot.sendMessage(chatId, '⚠️ Вы уже использовали свою способность.');
    showMainMenu(chatId);
    return;
  }

  user.energy -= 1;
  switch (user.character) {
    case 'Лумик':
      // Перемещение на две локации
      bot.sendMessage(chatId, '🌟 Вы можете переместиться на две локации за один ход!');
      // Логика перемещения на две локации
      chooseLocation(chatId, user);
      break;
    case 'Флаффи':
      // Повтор последнего действия
      bot.sendMessage(chatId, '🎉 Вы повторяете последнее действие!');
      // Логика повторения последнего действия
      search(chatId, user);
      break;
    case 'Тинори':
      // Просмотр следующего события
      const events = ['positive', 'negative', 'special'];
      const nextEvent = events[Math.floor(Math.random() * events.length)];
      bot.sendMessage(chatId, `🧠 Следующее событие при поиске будет: ${nextEvent}`);
      showMainMenu(chatId);
      break;
    case 'Ириска':
      if (!user.abilityUsed) {
        user.abilityUsed = true;
        user.markers += 1;
        bot.sendMessage(chatId, '🌈 Вы превратили негативное событие в позитивное и получили маркер!');
        checkVictoryCondition(chatId, user);
      } else {
        bot.sendMessage(chatId, '⚠️ Вы уже использовали свою способность.');
        showMainMenu(chatId);
      }
      break;
    default:
      bot.sendMessage(chatId, '⚠️ У вашего персонажа нет специальной способности.');
      showMainMenu(chatId);
      break;
  }
}

// Функция продвижения Тенехода
function advanceTenehod(chatId, user) {
  user.tenehodPosition += 1;
  if (user.tenehodPosition >= MAX_TENEHOD_POSITION) {
    bot.sendMessage(chatId, '🌑 Тенеход достиг дерева Луми. Игра окончена. Вы проиграли.');
  } else {
    bot.sendMessage(chatId, `🌑 Тенеход продвинулся! Текущее положение: ${user.tenehodPosition}`);
    showMainMenu(chatId);
  }
}

// Функция проверки условия победы
function checkVictoryCondition(chatId, user) {
  let totalMarkers = user.markers;
  if (user.specialItems.includes('✨ Волшебный артефакт')) {
    totalMarkers += 1; // Увеличиваем количество маркеров за счёт артефакта
  }
  if (totalMarkers >= REQUIRED_MARKERS) {
    bot.sendMessage(chatId, '🎉 Поздравляем! Вы собрали достаточное количество маркеров и спасли лес!');
  } else {
    showMainMenu(chatId);
  }
}

// Функция отображения статуса игрока
function showStatus(chatId, user) {
  const statusMessage = `
Персонаж: ${user.character}
Локация: ${user.location}
Маркеры 'Света Дружбы': ${user.markers}
Позиция Тенехода: ${user.tenehodPosition}
Энергия: ${user.energy}
Специальные предметы: ${user.specialItems.join(', ') || 'Нет'}
  `;
  bot.sendMessage(chatId, statusMessage);
  showMainMenu(chatId);
}

// Обработчик команды /help
bot.onText(/\/help/, (msg) => {
  const chatId = msg.chat.id;
  bot.sendMessage(chatId, 'Доступные команды:\n/start - начать игру\n/help - помощь');
});

// Установка команд бота
bot.setMyCommands([
  { command: '/start', description: 'Начать игру' },
  { command: '/help', description: 'Помощь по командам' }
]);

File: Kulenechki/Game/bot.py
import logging
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Updater, CommandHandler, CallbackContext, CallbackQueryHandler

logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)
logger = logging.getLogger(__name__)

REQUIRED_MARKERS = 5
MAX_TENEHOD_POSITION = 5

def start(update: Update, context: CallbackContext):
    user = update.message.from_user
    context.user_data['character'] = None
    context.user_data['location'] = 'Пушнор'
    context.user_data['markers'] = 0
    context.user_data['tenehod_position'] = 0
    keyboard = [
        [InlineKeyboardButton('Лумик', callback_data='Лумик')],
        [InlineKeyboardButton('Флаффи', callback_data='Флаффи')],
        [InlineKeyboardButton('Тинори', callback_data='Тинори')],
        [InlineKeyboardButton('Ириска', callback_data='Ириска')]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)
    update.message.reply_text(f"Привет, {user.first_name}! Выберите своего Куленёчка:", reply_markup=reply_markup)

def choose_character(update: Update, context: CallbackContext):
    query = update.callback_query
    character = query.data
    context.user_data['character'] = character
    query.edit_message_text(text=f"Вы выбрали {character}! Приключение начинается!")
    show_main_menu(query, context)

def show_main_menu(update_or_query, context):
    keyboard = [
        [InlineKeyboardButton('/move - переместиться на другую локацию', callback_data='move')],
        [InlineKeyboardButton("/search - искать маркеры 'Света Дружбы'", callback_data='search')],
        [InlineKeyboardButton('/status - показать текущий статус', callback_data='status')]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)
    if isinstance(update_or_query, Update):
        update_or_query.message.reply_text("Выберите действие:", reply_markup=reply_markup)
    else:
        update_or_query.message.reply_text("Выберите действие:", reply_markup=reply_markup)

def main_menu_handler(update: Update, context: CallbackContext):
    query = update.callback_query
    action = query.data
    if action == 'move':
        choose_location(query, context)
    elif action == 'search':
        search(query, context)
    elif action == 'status':
        status(query, context)

def choose_location(query, context):
    current_location = context.user_data.get('location', 'Пушнор')
    locations = {
        'Пушнор': ['Светящаяся поляна', 'Туманная роща'],
        'Светящаяся поляна': ['Пушнор', 'Ручей Забытых Снов'],
        'Туманная роща': ['Пушнор'],
        'Ручей Забытых Снов': ['Светящаяся поляна']
    }
    available_locations = locations.get(current_location, [])
    if available_locations:
        keyboard = [[InlineKeyboardButton(loc, callback_data=f'location_{loc}')] for loc in available_locations]
        reply_markup = InlineKeyboardMarkup(keyboard)
        query.edit_message_text("Куда вы хотите переместиться?", reply_markup=reply_markup)
    else:
        query.edit_message_text("Нет доступных локаций для перемещения.")
        show_main_menu(query, context)

def location_chosen(update: Update, context: CallbackContext):
    query = update.callback_query
    data = query.data
    if data.startswith('location_'):
        new_location = data.replace('location_', '')
        context.user_data['location'] = new_location
        query.edit_message_text(f"Вы переместились в {new_location}.")
        show_main_menu(query, context)

def search(query, context):
    import random
    events = ['positive', 'negative', 'special']
    event = random.choice(events)
    if event == 'positive':
        context.user_data['markers'] = context.user_data.get('markers', 0) + 1
        query.edit_message_text("Вы нашли маркер 'Света Дружбы'!")
        check_victory_condition(query, context)
    elif event == 'negative':
        advance_tenehod(query, context)
    else:
        query.edit_message_text("Вы встретили лесное существо, которое предлагает вам помощь.")
        show_main_menu(query, context)

def advance_tenehod(query, context):
    tenehod_position = context.user_data.get('tenehod_position', 0) + 1
    context.user_data['tenehod_position'] = tenehod_position
    if tenehod_position >= MAX_TENEHOD_POSITION:
        query.edit_message_text("Тенеход достиг дерева Луми. Игра окончена. Вы проиграли.")
    else:
        query.edit_message_text(f"Тенеход продвинулся! Текущее положение: {tenehod_position}")
        show_main_menu(query, context)

def check_victory_condition(query, context):
    markers = context.user_data.get('markers', 0)
    if markers >= REQUIRED_MARKERS:
        query.edit_message_text("Поздравляем! Вы собрали достаточное количество маркеров и спасли лес!")
    else:
        show_main_menu(query, context)

def status(query, context):
    character = context.user_data.get('character', 'Не выбран')
    location = context.user_data.get('location', 'Неизвестно')
    markers = context.user_data.get('markers', 0)
    tenehod_position = context.user_data.get('tenehod_position', 0)
    status_message = (
        f"Персонаж: {character}\n"
        f"Локация: {location}\n"
        f"Маркеры 'Света Дружбы': {markers}\n"
        f"Позиция Тенехода: {tenehod_position}"
    )
    query.edit_message_text(status_message)
    show_main_menu(query, context)

def help_command(update: Update, context: CallbackContext):
    update.message.reply_text(
        "Доступные команды:\n"
        "/start - начать игру\n"
        "/help - помощь"
    )

def main():
    token = '7314783664:AAEkBebeoJUKZFwcTsFP_rVt6resccsdYWs'
    updater = Updater(token)
    dp = updater.dispatcher
    dp.add_handler(CommandHandler('start', start))
    dp.add_handler(CommandHandler('help', help_command))
    dp.add_handler(CallbackQueryHandler(choose_character, pattern='^(Лумик|Флаффи|Тинори|Ириска)$'))
    dp.add_handler(CallbackQueryHandler(main_menu_handler, pattern='^(move|search|status)$'))
    dp.add_handler(CallbackQueryHandler(location_chosen, pattern='^location_'))
    updater.start_polling()
    updater.idle()

if __name__ == '__main__':
    main()
